### PR TITLE
Implement Invoke-SCuBADiff to diff two ScubaResults.json runs

### LINKED ISSUE #1210: Ability to produce "diff" between two assessment runs
# 💡 Summary #

ScubaGear produces assessment results and reports that indicate an M365 tenant's state at a given moment in time.  

## Motivation and context ##

The ability to see the change in security posture over time by comparing the results of two assessments is helpful to gauge progress towards or drift away from M365 SCB policies over time.

## Implementation notes ##

Please provide details for implementation, such as:

- Test the feasibility of comparing output results at different times
- Determine how to represent and format a "diff" between assessments
- Write code to do the diff and emit a diff result
- Develop an HTML diff report template

## Acceptance criteria ##

The following issues are completed

- [ ] create diff function to compare two scubaresults.json runs 
- [ ] identify and share precedence order
- [ ] diff function will export .json, .html, and .csv files
- [ ] diff function will take into consideration the securitysuite migration policies
- [ ] new markdown created to detail diff feature and relevant information

### LINKED ISSUE #2414: Implement Invoke-SCuBADiff, the validated diff command for two ScubaResults.json runs
### Prerequisites

- [x] This issue has an informative and human-readable title.
- [x] Assignee should meet with issue author before starting work.

### 💡 Summary

Follow-up to #1210, which was the investigation and prototyping issue. This issue is the result of that investigation: implement and ship Invoke-SCuBADiff, the validated command that compares a "before" and "after" ScubaResults.json and reports how every policy's result changed between the two runs, exactly as specified on the ScubaGearDiffproto branch (docs/execution/diff.md) after design review and team feedback at the August 2026 SCuBA TEM and several PR happy hours.

### Motivation and context

#1210 scoped and answered the open questions: whether comparing two assessment runs was feasible, how to represent the diff, and whether a working prototype held up. It did - Invoke-SCuBADiff was demoed at the August 2026 SCuBA TEM Diff Tool session, the team gave feedback on the design (see the session notes for the questions raised, e.g. how a Defender-to-Security-Suite migration should be handled), and the ScubaGearDiffproto branch reflects that validated design. This issue is not another investigation - it's implementing that already-agreed design as a shippable ScubaGear command, merging it out of the prototype branch and into mainline. This directly supports BOD 25-01 continuous reporting, since tracking change in security posture over time is the whole point of the diff.

### Implementation notes

1. Command surface - implement Invoke-SCuBADiff -BeforePath <path> -AfterPath <path>, with -OutPath (folder, created if missing), -OutJsonFileName, -OutCsvFileName, -OutReportFileName, and -DarkMode (defaults the HTML report to dark theme), as specified. Returns an object with JsonPath, CsvPath, and ReportPath.
2. Base-ID / version-suffix matching - implement matching on base ID (trailing vN suffix stripped, e.g. MS.AAD.1.1v1 -> MS.AAD.1.1). Same base ID + same version compares directly; same base ID + different version classifies as PolicyVersionUpdate (informational only); a base ID present in only one file is NewPolicy or RemovedPolicy.
3. Security Suite migration feature - implement the fixed Defender/EXO -> Security Suite migration table (13 Defender rows, 3 EXO rows) as specified, derived from PowerShell/ScubaGear/mappings/scuba-baseline-policy-migrations.csv restricted to one-to-one rows. A migrated pair files under Security Suite with both control IDs shown (e.g. MS.DEFENDER.4.1v2 -> MS.SECURITYSUITE.3.1v1) and a "migrated" badge; the retired policy also stays in its original product's section as a greyed-out Migrated record with blank after-columns, pointing at its replacement via MigratedToId/MigratedToProduct. A direct base-ID match always wins over the migration alias, and the alias is directional. Not migrated, per spec, and left to fall through to normal RemovedPolicy handling: MS.DEFENDER.1.1v1-1.5v1, MS.DEFENDER.4.5v1, the MS.EXO.14 and MS.EXO.15 groups, and the remaining EXO/Teams rows in the mapping CSV. Generate or derive the migration table from scuba-baseline-policy-migrations.csv directly (not hand-copied), so it can't silently drift as that CSV changes.
4. Classification taxonomy and precedence order - implement the full 9-rank precedence order as specified (0 Migrated, 1 NewPolicy/RemovedPolicy, 2 Errored, 3 PolicyVersionUpdate, 4 Unchanged, 5 NewIncorrectResult, 6 the specific-diff set [NewFail, NewPass, NewWarning, NewAutomatedCheck, NewManualCheck], 7 NewOmission, 8 Other), and the full before-to-after diff table from the design doc as the canonical classification reference. A control matching more than one rule resolves to the first matching rank. Result is an open string set - an unrecognized value classifies as Other with both literal values preserved rather than crashing.
5. Output artifacts - implement DiffResults.json (schema-versioned via a top-level SchemaVersion field, nested by product), DiffResults.csv (flattened to one row per policy with a leading Product column, includes Unchanged rows, fixed column list per the design doc), and DiffReport.html (self-contained, color-coded, Unchanged rows hidden by default). Every CSV row carries every column, with fields left blank where not applicable (false-positive fields, annotation fields, Migrated-From/To fields, before/after fields for NewPolicy/RemovedPolicy/Migrated rows).
6. HTML filters - implement one filter checkbox per classification in the summary table, including classifications with zero rows in the current diff; unchecking a classification hides its rows in the per-product tables, dims its summary column, and recomputes each product's Total. Unchanged is filtered separately via a "Show unchanged rows" toggle and is always counted in the Total. Dark Mode toggle, defaulting from -DarkMode.
7. Row coloring - rows colored by Result (After), not by diff type: red for Fail/Error, yellow for Warning, green for Pass, grey for N/A/Omitted/other and for RemovedPolicy rows.
8. Annotations (Fail -> Fail) - for controls failing in both runs, compare AnnotatedFailedPolicies entries and add AnnotationChanged, Comment, and RemediationDate fields, scoped to Fail -> Fail records only, per spec.
9. False positives (Incorrect result) - recognize ScubaGear's "Incorrect result" marking and classify the change in the marking itself (NewIncorrectResult when newly marked; the revealed result's classification when a false-positive marking is removed; Unchanged when stable). Add MarkedIncorrectBefore/After and UnderlyingResultBefore/After (the tool-computed OriginalResult).
10. Markdown documentation - move docs/execution/diff.md off the ScubaGearDiffproto branch into mainline docs as the shipped reference doc, verifying its relative links (to removedpolicies.md and scuba-baseline-policy-migrations.csv) resolve correctly in mainline's docs structure.
11. README reference - add Invoke-SCuBADiff to the top-level README (and/or the docs index alongside Invoke-SCuBA) so the diff command is discoverable without already knowing docs/execution/diff.md exists.
12. Items still open after the TEM session - these were raised as discussion, not resolved as part of the validated design, and are out of scope for this issue: (a) users onboarded to ScubaConnect cannot use the Diff tool today since they don't have access to their own result files outside CDM - a known limitation to document, not solve here; (b) multi-report comparison and trend analysis remain out of scope for v1 (this command compares exactly two reports), deferred pending agency feedback and given CDM already offers some trend analysis; (c) auto-sending diff results to ScubaConnect-onboarded users and adding a points-of-contact field to the config file remain undecided and are not built here. Extending the diff beyond policy results to AAD report tables (risky apps, risky service principals, Conditional Access policies, etc.) is being scoped separately - see the companion issue.
13. Testing - unit tests covering the precedence order (especially "direct match wins over migration alias" and the deliberately-not-migrated carve-outs) and the full diff table's before/after combinations; functional tests running Invoke-SCuBADiff against representative before/after ScubaResults.json fixtures, including a fixture pair that exercises the Security Suite migration.

### Acceptance criteria

- [ ] Invoke-SCuBADiff implemented with -BeforePath, -AfterPath, -OutPath, -OutJsonFileName, -OutCsvFileName, -OutReportFileName, and -DarkMode as specified
- [ ] Base-ID and version-suffix matching implemented, including PolicyVersionUpdate as an informational-only classification
- [ ] Security Suite migration feature implemented: migration table derived from scuba-baseline-policy-migrations.csv, Migrated classification and badge, direct-match-wins rule, directional alias, and the specified non-migrated carve-outs
- [ ] Full classification taxonomy implemented with the 9-rank precedence order enforced
- [ ] Before-to-after diff table implemented as the canonical classification logic and documented as such
- [ ] Diff function exports DiffResults.json, DiffResults.csv, and DiffReport.html exactly as specified
- [ ] HTML report has a filter checkbox for every classification (including zero-row ones), a separate "Show unchanged rows" toggle, and a Dark Mode toggle
- [ ] Row coloring by Result (After), Fail->Fail annotation tracking, and false-positive/Incorrect-result handling implemented per spec
- [ ] docs/execution/diff.md merged into mainline docs (off the ScubaGearDiffproto branch), with internal links verified
- [ ] Top-level README (and/or docs index) updated to reference the new diff command
- [ ] Still-open items (ScubaConnect accessibility gap, multi-report/trend analysis, points-of-contact field) documented in Notes as explicitly out of scope for this issue
- [ ] Unit tests cover the classification precedence order and migration edge cases; functional tests run against representative before/after ScubaResults.json fixtures

### PR DESCRIPTION
**Pull request by ahuynhECS**
## 🗣 Description ##
 
Implements `Invoke-SCuBADiff`, a new offline command that compares a "before" and "after" `ScubaResults.json` and reports how every policy's result changed between the two runs. This is the validated design from the `ScubaGearDiffproto` branch (`docs/execution/diff.md`), demoed and reviewed with the team at the August 2026 SCuBA TEM Diff Tool session - this PR implements that already-agreed design rather than re-designing it.
 
Included in this PR, all in service of the single goal of shipping the diff command:
 
- `Invoke-SCuBADiff -BeforePath -AfterPath`, with `-OutPath`, `-OutJsonFileName`, `-OutCsvFileName`, `-OutReportFileName`, and  `-DarkMode`
- Base-ID / version-suffix matching, including `PolicyVersionUpdate`
- Security Suite migration table (Defender/EXO -> Security Suite) with the `Migrated` classification and badge
- Full 9-rank classification taxonomy and precedence order
- Three output artifacts: `DiffResults.json`, `DiffResults.csv`, `DiffReport.html`
- HTML report: per-classification filters, "Show unchanged rows" toggle, Dark Mode
- Row coloring by Result (After), Fail -> Fail annotation tracking, and Incorrect-result / false-positive handling
- `docs/execution/diff.md` merged into mainline docs; README updated to reference the new command
- Unit and functional test coverage
 
## 💭 Motivation and context ##
 
Closes #2414 
 
Agencies need a way to track "change in security posture over time" between two ScubaGear runs, which directly supports BOD 25-01 continuous reporting. Without this, comparing two reports is a manual, error-prone exercise of reading two full HTML reports side by side.
 
## 🧪 Testing ##
 
**Testing checklist**
 
- [ ] **Command surface.** Run `Invoke-SCuBADiff -BeforePath <before> -AfterPath <after> -OutPath <folder> -DarkMode`. Confirm it writes `DiffResults.json`, `DiffResults.csv`, and `DiffReport.html` (or your custom `-OutJsonFileName`/`-OutCsvFileName`/`-OutReportFileName` names) to `<folder>`, and that the returned object exposes `JsonPath`, `CsvPath`, and `ReportPath` pointing at them.
- [ ] **Base-ID / version matching, `PolicyVersionUpdate`.** Not present in the attached fixture - build a small before/after pair with the same base ID at two different `vN` suffixes (e.g. a `v1` before, `v2` after) and confirm it classifies as `PolicyVersionUpdate`, is treated as informational only (not counted as an authoritative pass/fail delta), and both results still render.
- [ ] **Security Suite migration.** In the attached output, confirm: `MS.DEFENDER.2.1v1`-`MS.DEFENDER.6.3v1` and the retired DLP/alert/audit policies appear once in the Defender section as plain `Migrated` rows with blank after-columns. Confirm the migration table is generated from (or checked against) `scuba-baseline-policy-migrations.csv`, not a hand-maintained copy that can drift.
- [ ] **Full classification taxonomy + precedence.** Cross-check `DiffResults.csv`'s `Classification` column against the table above - all 10 represented classifications should match counts. Spot-check `MS.AAD.7.1v1` (`Pass` -> `Incorrect result`, newly marked -> `NewIncorrectResult`) and `MS.AAD.3.1v1` (previously marked incorrect, now `Fail` -> `NewFail`, revealing the underlying result) to confirm precedence resolves correctly when more than one rule could apply (`[screenshot: 02-aad-classifications.png]`). For `NewAutomatedCheck`/`NewManualCheck`/`Other`, use synthetic pairs as noted above.
- [ ] **Diff table as canonical reference.** Confirm the before-to-after diff table in `docs/execution/diff.md` matches the actual classification logic in code line for line - a documentation-accuracy check, not just a behavioral one.
- [ ] **Three output artifacts.** Open all three: confirm `DiffResults.json` carries a top-level `SchemaVersion` (this sample is `1.0`) and is nested by product; confirm `DiffResults.csv` is flattened to one row per policy with a leading `Product` column and **includes** `Unchanged` rows (64 of them in this sample - the HTML hides these by default, the CSV should not); confirm every CSV row has all 24 columns present, blank where not applicable (e.g. `MigratedFromId` only populated for `Migrated` rows).
- [ ] **HTML filters, Show unchanged rows, Dark Mode.** Open `DiffReport.html`. Confirm the summary table has a filter checkbox for every classification, including the four with zero rows in this sample (`PolicyVersionUpdate`, `Other`, `NewAutomatedCheck`, `NewManualCheck`) (`[screenshot: 01-summary-and-legend.png]`). Uncheck a couple of classifications and confirm their rows disappear from the per-product tables, their summary column dims, and each product's Total recomputes (`[screenshot: 07-classification-filters.png]`). Toggle "Show unchanged rows" and confirm the 64 `Unchanged` rows appear/disappear correctly (`[screenshot: 06-show-unchanged-rows.png]`). Toggle Dark Mode and confirm the whole report re-themes, including the legend and summary table (`[screenshot: 05-dark-mode.png]`).
- [ ] **Row coloring, annotations, false positives.** Confirm rows are colored by Result (After) - red/Fail, yellow/Warning, green/Pass, grey/N/A-Omitted-Removed - not by diff type. Confirm `MS.AAD.7.1v1`'s row shows `Incorrect result (underlying: Fail)` inline and is greyed out per the `NewIncorrectResult` rule (`[screenshot: 02-aad-classifications.png]`). For Fail -> Fail annotation tracking (`AnnotationChanged`/`Comment`/`RemediationDate`), this sample has no `AnnotationChanged: true` case - build a synthetic pair where a `Fail -> Fail` record's comment or remediation date differs and confirm those three fields populate correctly.
- [ ] **Documentation review - `docs/execution/diff.md`.** Read the shipped doc end to end against actual command behavior, not just against the old branch draft. Check: the command surface and parameter table match the real signature; the migration table and "what is deliberately not migrated" list match `scuba-baseline-policy-migrations.csv`; the precedence-order and diff tables match the classification logic above; the CSV column list matches the real output; internal relative links (`removedpolicies.md`, `scuba-baseline-policy-migrations.csv`) resolve now that this has moved into mainline docs.
- [ ] **README / docs index.** Confirm `Invoke-SCuBADiff` is discoverable from the top-level README (and/or docs index) without already knowing `docs/execution/diff.md` exists.
- [ ] **Test coverage.** Confirm unit tests exist for the precedence order (including "direct match wins over migration alias" and the non-migrated carve-outs) and for the full diff table's before/after combinations, and that functional tests run `Invoke-SCuBADiff` against fixtures similar to the attached pair.

 
## 📷 Screenshots ##
 
01-summary-and-legend
Summary table, legend, and before/after source cards at the top of the report.
 

02-aad-classifications
AAD product table showing `NewFail`, `NewPass`, `NewPolicy`, `NewIncorrectResult` (with the `(underlying: Fail)` tooltip), and `NewOmission` rows.
 

03-removed-and-migrated-source-side
Defender product section: `RemovedPolicy` rows and the source-side `Migrated` rows with blank after-columns.
 

04-migrated-badge-destination-side
Security Suite section: the destination-side rows carrying the **migrated** badge (e.g. `MS.DEFENDER.4.1v2 -> MS.SECURITYSUITE.3.1v1`), plus `NewPolicy` and `Errored` examples.
 

05-dark-mode
Dark Mode toggle applied to the summary table.
 
06-show-unchanged-rows
"Show unchanged rows" toggle revealing the 64 `Unchanged` rows in the AAD section.
 
07-classification-filters
`NewFail` and `NewPass` classification filters unchecked, showing their rows hidden from the AAD table and the summary column dimmed.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [x] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [ ] PR passed smoke test check.
- [ ] PR/feature branch passes functional tests for relevant products, if applicable.
- [ ] Feature branch has been rebased against changes from parent branch, as needed.

  Use `Update branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [ ] Resolved all merge conflicts on branch.
- [ ] Squash all commits into one PR level commit using the `Squash and merge` button.

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Close issues resolved by this PR if the [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) did not activate.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

**Comment by FollyBeachGurl** [comment]
Suggested we add one additional reviewer because of how impactful this PR is. Will revisit next meeting.

**Review by sylvesterkaczmarek — COMMENTED** [review]
These writes use Set-Content -Path even though the paths come from user-supplied OutPath/file names. A valid path containing [ or ] is treated as a wildcard and can fail or hit the wrong target. Can these use -LiteralPath like the input reader, with a bracket-path regression?

File: PowerShell/ScubaGear/Modules/Diff/Diff.psm1 [hunk 1]
```diff
@@ -0,0 +1,1487 @@
...
+                'CriticalityBefore'      = $r.CriticalityBefore
+                'CriticalityAfter'       = $r.CriticalityAfter
+                'Requirement'            = $r.Requirement
+                'DetailsAfter'           = $r.DetailsAfter
+                'MarkedIncorrectBefore'  = & $get 'MarkedIncorrectBefore'
+                'MarkedIncorrectAfter'   = & $get 'MarkedIncorrectAfter'
+                'UnderlyingResultBefore' = & $get 'UnderlyingResultBefore'
+                'UnderlyingResultAfter'  = & $get 'UnderlyingResultAfter'
+                'AnnotationChanged'      = & $get 'AnnotationChanged'
+                'Comment'                = & $get 'Comment'
```
Comment (skirkpatrickMSFT) [inline PowerShell/ScubaGear/Modules/Diff/Diff.psm1]: **CSV formula injection (minor).** `Comment` is free text an analyst types into `AnnotatedFailedPolicies`, and it flows into `DiffResults.csv` unsanitized. A comment starting with `=`, `+`, `-`, or `@` could be interpreted as a formula if the CSV is later opened in Excel/Sheets by someone else. I confirmed the existing `ConvertTo-ResultsCsv` path has the same gap, so this isn't a regression, but since `Comment` is genuinely human-authored (higher risk than the mostly baseline-controlled fields around it), it may be worth prefixing a leading `'` (or similar) on write here, or tracking as a follow-up issue rather than carrying it forward silently.

File: PowerShell/ScubaGear/Modules/Diff/Diff.psm1 [hunk 2]
```diff
@@ -0,0 +1,1487 @@
...
+    # comparison is informational, so this outranks the specific diffs. Skipped for
+    # a migrated pair: the two IDs come from different policy families (e.g.
+    # MS.DEFENDER.4.1v2 -> MS.SECURITYSUITE.3.1v1), so their version suffixes are
+    # not comparable and the pair is classified by its result like any other.
+    if (-not $IsMigrated -and $BeforeVersion -ne $AfterVersion) { return 'PolicyVersionUpdate' }
+
+    # Identical Result and version. Checked before the diffs below so a
+    # stable state -- including a stable "Incorrect result" marking or a control
+    # omitted in both runs -- is never mistaken for a diff.
+    if ($bCat -eq $aCat -and $BeforeResult -eq $AfterResult) { return 'Unchanged' }
```
Comment (skirkpatrickMSFT) [inline PowerShell/ScubaGear/Modules/Diff/Diff.psm1]: **Masked false-positive drift is invisible.** When both sides carry the literal `"Incorrect result"` marking, this line classifies the pair `Unchanged` based only on the literal Result string -- it never compares `OriginalResult` (the `UnderlyingResultBefore`/`After` fields populated later are informational only, not part of classification). So if an analyst re-marks a worsening result as incorrect in both runs (e.g. underlying Fail -> Warning while staying masked), the diff reports `Unchanged` and that drift is only visible if someone manually cross-references the underlying-result columns. Test coverage (`Treats a stable false-positive marking as Unchanged`) only exercises the case where the underlying result is identical on both sides. Is silently swallowing a changed *underlying* result here intentional for v1, or worth a follow-up?

File: PowerShell/ScubaGear/Modules/Diff/Diff.psm1 (hunk 1 above)
Comment (skirkpatrickMSFT) [inline PowerShell/ScubaGear/Modules/Diff/Diff.psm1:1070]: Comment here is free text an analyst types when annotating a failed policy, and it goes straight into the CSV with no escaping. If that value starts with `=`, `+`, `-`, or `@`, Excel (and most other spreadsheet apps) will try to evaluate it as a formula when the CSV is opened -- could be something as simple as a `HYPERLINK()` that phones home, or worse on older Excel builds with DDE still enabled. Whoever opens this CSV (probably not the person who ran the diff, more likely a compliance reviewer down the line per the BOD 25-01 use case) is the one at risk.

Not saying this should block the PR -- the existing CSV export in `Orchestrator.psm1` has the same gap -- but since `Comment` is the one genuinely free-text field going into this file, might be worth prefixing a leading `'` on anything starting with `=`/`+`/`-`/`@` before writing it out. Could also just be a follow-up issue if you'd rather not scope-creep this PR.

File: PowerShell/ScubaGear/Modules/Diff/Diff.psm1 (hunk 2 above)
Comment (skirkpatrickMSFT) [inline PowerShell/ScubaGear/Modules/Diff/Diff.psm1:390]: This line marks a pair `Unchanged` any time both sides are literally `"Incorrect result"` -- but it's only comparing the literal `Result` string, never `OriginalResult`. So if someone re-marks a control as a false positive in both runs, but the real result underneath got worse (say `Warning` before, `Fail` after), this reports `Unchanged`. It's hidden from the report by default, and even with "show unchanged rows" on, it just renders as a plain grey row -- nothing flags that the underlying result changed, even though `UnderlyingResultBefore`/`After` are sitting right there in the data.

Given the whole point of this tool is catching security posture changes over time, this feels like exactly the case it should catch -- a control quietly getting worse while masked by a stable annotation. Is that intentional for v1 (similar to how the `Fail -> Fail` annotation comparison is deliberately narrow), or worth a fix before this ships?

**Review by skirkpatrickMSFT — COMMENTED** [review]
Nice work on this one -- solid test coverage on the classification precedence table, and the docs actually match the code (rare!). Left two inline notes, neither should block merging.

**Comment by DickTracyII** [comment]
> These writes use Set-Content -Path even though the paths come from user-supplied OutPath/file names. A valid path containing [ or ] is treated as a wildcard and can fail or hit the wrong target. Can these use -LiteralPath like the input reader, with a bracket-path regression?

@ahuynhECS 
Funny because I was going to say the same. I had a folder with PR testing in called `PR2416 [Test]` and it could not find the file unless I used single quotes for literal path. An option is to use `-LiteralPath` instead of `-Path` to the `-BerforePath` and `-AfterPath` parameters.

I'd also say we should add these files to the gitignore in case someone (like me) ran it without the `-Outpath` and it created it the Branch folder.

Great tool BTW. My only input is when I first read it, it makes since but my eye did not catch that I passed before and failed after. Not sure the best way to show this, maybe use a dark green for pass and dark red for fail? also should there be darker lines for the tables like scuba results show?
image

File: PowerShell/ScubaGear/Modules/Diff/Diff.psm1 [hunk 3]
```diff
@@ -0,0 +1,1487 @@
...
+#     groups are excluded together so the reworked anti-spam and Safe Links
+#     policies are read on their own terms.
+#   - Every remaining EXO row and every Teams row. Most collapse many-to-one onto
+#     a target a Defender row already claims (MS.EXO.16.1v1 and MS.DEFENDER.5.1v1
+#     both map to MS.SECURITYSUITE.4.1v1, and MS.EXO.11.1v1 maps to
+#     MS.SECURITYSUITE.2.1v1, which MS.DEFENDER.2.1v1 owns); the rest are either
+#     retired outright (New ID "None", e.g. MS.EXO.11.3v1) or manual checks that
+#     only ever produced N/A, so an aligned before/after comparison would carry no
+#     signal.
+$script:PolicyMigrationMap = [ordered]@{
```
Comment (DickTracyII) [inline PowerShell/ScubaGear/Modules/Diff/Diff.psm1]: could this be mapped to the scubabaseline.json. Like what New-scubaconfig and Start-scubaconfigapp uses. This would make it more dynamic incase basleline ever change or update

File: PowerShell/ScubaGear/Modules/Diff/Diff.psm1 [hunk 4]
```diff
@@ -0,0 +1,1487 @@
...
+        [ValidateNotNullOrEmpty()]
+        [string]
+        $OutReportFileName = 'DiffReport',
+
+        [Parameter(Mandatory = $false)]
+        [switch]
+        $DarkMode
+    )
+
+    $Before = Import-ScubaResultsFile -Path $BeforePath
```
Comment (DickTracyII) [inline PowerShell/ScubaGear/Modules/Diff/Diff.psm1]: Change these to `-LiteralPath` to support spaces and weird supported characters as folders. this may have be changed in several functions too

File: PowerShell/ScubaGear/Modules/Diff/Diff.psm1 (hunk 1 above)
Comment (ahuynhECS) [reply PowerShell/ScubaGear/Modules/Diff/Diff.psm1]: will make the updates!

**Comment by ahuynhECS** [comment]
> > These writes use Set-Content -Path even though the paths come from user-supplied OutPath/file names. A valid path containing [ or ] is treated as a wildcard and can fail or hit the wrong target. Can these use -LiteralPath like the input reader, with a bracket-path regression?
> 
> @ahuynhECS Funny because I was going to say the same. I had a folder with PR testing in called `PR2416 [Test]` and it could not find the file unless I used single quotes for literal path. An option is to use `-LiteralPath` instead of `-Path` to the `-BerforePath` and `-AfterPath` parameters.
> 
> I'd also say we should add these files to the gitignore in case someone (like me) ran it without the `-Outpath` and it created it the Branch folder.
> 
> Great tool BTW. My only input is when I first read it, it makes since but my eye did not catch that I passed before and failed after. Not sure the best way to show this, maybe use a dark green for pass and dark red for fail? also should there be darker lines for the tables like scuba results show? image

updated to -literalpath and added to gitignore!

**Comment by ahuynhECS** [comment]
> > > These writes use Set-Content -Path even though the paths come from user-supplied OutPath/file names. A valid path containing [ or ] is treated as a wildcard and can fail or hit the wrong target. Can these use -LiteralPath like the input reader, with a bracket-path regression?
> > 
> > 
> > @ahuynhECS Funny because I was going to say the same. I had a folder with PR testing in called `PR2416 [Test]` and it could not find the file unless I used single quotes for literal path. An option is to use `-LiteralPath` instead of `-Path` to the `-BerforePath` and `-AfterPath` parameters.
> > I'd also say we should add these files to the gitignore in case someone (like me) ran it without the `-Outpath` and it created it the Branch folder.
> > Great tool BTW. My only input is when I first read it, it makes since but my eye did not catch that I passed before and failed after. Not sure the best way to show this, maybe use a dark green for pass and dark red for fail? also should there be darker lines for the tables like scuba results show? image
> 
> updated to -literalpath and added to gitignore!

also just updated the colors!

File: PowerShell/ScubaGear/Modules/Diff/Diff.psm1 (hunk 2 above)
Comment (ahuynhECS) [reply PowerShell/ScubaGear/Modules/Diff/Diff.psm1]: hmm for this one i had kept it grey under the assumption that the agency had reviewed the result from both runs and made the comment to mark it as incorrect so it is "unchanged / doesn't need further review" from their point of view

Comment (ahuynhECS) [reply PowerShell/ScubaGear/Modules/Diff/Diff.psm1]: also for the example of Warning before into Fail after this should be a change in criticality from should to shall so it should trigger a policy version bump so it should be picked up under newpolicyupdate for the diff.

**Comment by FollyBeachGurl** [comment]
potential rollover

**Review by skirkpatrickMSFT — APPROVED** [review]
The changes I requested are in place. Good job!

**Comment by ahuynhECS** [comment]
@DickTracyII good for you to re-review
